### PR TITLE
Add IPC FileDecoder

### LINKED ISSUE #5153: Request to Memmap Arrow IPC files on disk
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I have machine learning dataset that exceed my system's ram which I access in a random pattern. I currently split the data into a flat f32 array written to the disk, and a set of parquet files. This is awkward and requires complex glue code. 

**Describe the solution you'd like**
I'd like to use a memmap like this: https://huggingface.co/docs/datasets/about_arrow

**Describe alternatives you've considered**
Switching to another format. rkvy works, but is not nearly as interoperable. 

**Additional context**

### LINKED ISSUE #5252: Support get offsets or blocks info from arrow file.
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

It seems that there isn't a pub function to provide offsets or blocks info from arrow file. 

**Describe the solution you'd like**

In arrow-ipc/src/reader.rs, 
```rust
impl<R: Read + Seek> FileReader<R> {
    pub fn blocks(&self) -> Vec<Block> {
        &self.blocks
    }
   //OR
    pub fn offsets(&self) -> Vec<i64> {
        &self.blocks.iter().map(Block::offset).collect()
    }
}
```

**Describe alternatives you've considered**

**Additional context**

related to https://github.com/apache/arrow-datafusion/issues/8503

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #5153
Closes #5165
Closes #5252
Part of #1207

# Rationale for this change
 

In a similar vein to arrow-csv and arrow-json this extracts a low-level, push-based interface for decoding files. This lends itself to use-cases that need more control over how IO is performed, for example:

* Async IO
* Memory mapped Buffer
* Parallel Decode

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
@comath perhaps you might take a look, whilst perhaps not as simple as https://github.com/apache/arrow-rs/pull/5165 I think this is a much more flexible and therefore future-proof mechanism to achieve this

File: arrow-ipc/src/reader.rs [hunk 1]
```diff
@@ -522,6 +522,174 @@ fn parse_message(buf: &[u8]) -> Result<Message, ArrowError> {
...
+/// }
+///
+/// // Read record batch
+/// let batches = footer.recordBatches().unwrap();
+/// assert_eq!(batches.len(), 1); // Only wrote a single batch
+///
+/// let block = batches.get(0);
+/// let block_len = block.bodyLength() as usize + block.metaDataLength() as usize;
+/// let data = buffer.slice_with_length(block.offset() as _, block_len);
+/// let back = decoder.read_record_batch(block, &data).unwrap().unwrap();
```
Comment (tustvold) [inline arrow-ipc/src/reader.rs:598]: This is still somewhat verbose, but I couldn't see an easy way to reduce this further that didn't end up in knots with self-referential structs (as flatbuffers borrow data)

File: arrow-ipc/src/reader.rs [hunk 2]
```diff
@@ -522,6 +522,174 @@ fn parse_message(buf: &[u8]) -> Result<Message, ArrowError> {
...
+                "Expecting DictionaryBatch in dictionary blocks, found {t:?}."
+            ))),
+        }
+    }
+
+    /// Read the RecordBatch with the given block and data buffer
+    pub fn read_record_batch(
+        &self,
+        block: &Block,
+        buf: &Buffer,
```
Comment (tustvold) [inline arrow-ipc/src/reader.rs:663]: It is perhaps worth noting that this interface won't allow pushing down column projection to IO, I think this is a bridge we cross when we add support for this.

File: arrow-ipc/src/reader.rs [hunk 3]
```diff
@@ -522,6 +522,174 @@ fn parse_message(buf: &[u8]) -> Result<Message, ArrowError> {
...
+            return Err(ArrowError::IpcError(
+                "Could not read IPC message as metadata versions mismatch".to_string(),
+            ));
+        }
+        Ok(message)
+    }
+
+    /// Read the dictionary with the given block and data buffer
+    pub fn read_dictionary(&mut self, block: &Block, buf: &Buffer) -> Result<(), ArrowError> {
+        let message = self.read_message(buf)?;
```
Comment (viirya) [inline arrow-ipc/src/reader.rs:641]: Not bit deal but seems we didn't check for metadata version for `DictionaryBatch` before. Maybe it was missed before.

**Review by viirya — APPROVED** [review]

File: arrow-ipc/src/reader.rs [hunk 4]
```diff
@@ -522,6 +522,174 @@ fn parse_message(buf: &[u8]) -> Result<Message, ArrowError> {
         .map_err(|err| ArrowError::ParseError(format!("Unable to get root as message: {err:?}")))
 }
 
+/// Read the footer length from the last 10 bytes of a file
```
Comment (viirya) [inline arrow-ipc/src/reader.rs:525]:
```suggestion
/// Read the footer length from the last 10 bytes of an Arrow IPC file
```

**Comment by alamb** [comment]
Thank you for the reviews @viirya  and the code @tustvold

**Comment by comath** [comment]
> @comath perhaps you might take a look, whilst perhaps not as simple as #5165 I think this is a much more flexible and therefore future-proof mechanism to achieve this

I'm a bit late, but I like it. It exposes the mechanisms for people to do their own crazy things in the future.

**Comment by alamb** [comment]
> It exposes the mechanisms for people to do their own crazy things in the future.

😆  -- love it!

**Pull request merged**